File: src/main/scala/ConsumerDemo.scala
package com.example


import java.text.SimpleDateFormat
import java.util.{Collections, Properties}

import org.apache.kafka.clients.consumer.{ConsumerConfig, KafkaConsumer}

import scala.collection.JavaConversions._
import org.apache.kafka.common.TopicPartition
import java.util

object ConsumerDemo {
  val dateFormat = "yyyy-MM-dd HH:mm:ss"
  val epochToDate: Long => String = new SimpleDateFormat(dateFormat).format(_)

  def main(args: Array[String]): Unit = {

    val props = new Properties()
    props.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, "localhost:9092")
    props.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, "org.apache.kafka.common.serialization.StringDeserializer")
    props.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, "org.apache.kafka.common.serialization.StringDeserializer")
    props.put(ConsumerConfig.GROUP_ID_CONFIG, "consumergroup1_rajesh")
    props.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "latest")
    val consumer = new KafkaConsumer[String, String](props)
    // consumer.subscribe(Collections.singletonList(envProps.getString("topic")))
    // consumer.subscribe(Collections.singletonList("rajeshs_t1"))
    val tp = new TopicPartition("Testtopic1", 0)
    val tps = util.Arrays.asList(tp)
    consumer.assign(tps)
    consumer.seekToBeginning(tps)
    while (true) {
      val records = consumer.poll(500)
      for (record <- records.iterator()) {
        println("received metadata : \n " +
          "Topic      : " + record.topic() + "\n" +
          "Partiton   : " + record.partition() + "\n" +
          "Offset     : " + record.offset() + "\n" +
          "TimeStamp  : " + record.timestamp() + "\n" +
          "Date  : " + epochToDate(record.timestamp()) + "\n" +
          "Value      : " + record.value())
      }
    }
  }
}

File: Kafka-basics/src/main/java/kafka/ConsumerDemo.java
package kafka;

import org.apache.kafka.clients.consumer.ConsumerConfig;
import org.apache.kafka.clients.consumer.ConsumerRecord;
import org.apache.kafka.clients.consumer.ConsumerRecords;
import org.apache.kafka.clients.consumer.KafkaConsumer;
import org.apache.kafka.common.serialization.StringDeserializer;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.time.Duration;
import java.util.Arrays;
import java.util.Properties;

public class ConsumerDemo {
    public static void main(String[] args) {

        Logger logger = LoggerFactory.getLogger(ConsumerDemo.class);
        String bootstrapServers = "localhost:9092";
        String topic = "FirstTopic1";
        String groupId = "my4thApp";

        Properties properties = new Properties();

        properties.setProperty(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, bootstrapServers);
        properties.setProperty(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class.getName());
        properties.setProperty(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class.getName());
        properties.setProperty(ConsumerConfig.GROUP_ID_CONFIG, groupId);
        properties.setProperty(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest"); // latest//none

        KafkaConsumer<String, String> consumer = new KafkaConsumer<String, String>(properties);
        consumer.subscribe(Arrays.asList(topic));

        while (true) {
            ConsumerRecords<String, String> records = consumer.poll(Duration.ofMillis(100)); // new in kafka 2.0
            for (ConsumerRecord<String, String> record : records) {
                logger.info("received metadata : \n " +
                        "Topic      : " + record.topic() + "\n" +
                        "Partiton   : " + record.partition() + "\n" +
                        "Offset     : " + record.offset() + "\n" +
                        "TimeStamp  : " + record.timestamp() + "\n" +
                        "Value      : " + record.value());
            }

        }

    }
}
